redaxo/src/core/lib/list.php

```diff
@@ -909,6 +922,8 @@ public function replaceVariable($string, $varname)
      * @param string $value Zu durchsuchender String
      *
      * @return string
+     *
+     * @psalm-taint-specialize
```

Comment:
wieso brauchen wir das? kann diese methode überhaupt bzgl. taint irgendwas escapen?

Reply:
Ich finde es auch merkwürdig. Ohne würde ich diesen Fehler bekommen:

```
ERROR: TaintedHtml - redaxo/src/core/lib/list.php:1156:14 - Detected tainted HTML (see https://psalm.dev/245)
  $_REQUEST
    <no known location>

  call to rex_request::arrayKeyCast - redaxo/src/core/lib/request.php:55:35
        return self::arrayKeyCast($_REQUEST, $varname, $vartype, $default);

  rex_request::arrayKeyCast#1 - redaxo/src/core/lib/request.php:204:48
    private static function arrayKeyCast(array $haystack, $needle, $vartype, $default = '')

  $haystack - redaxo/src/core/lib/request.php:204:48
    private static function arrayKeyCast(array $haystack, $needle, $vartype, $default = '')

  $haystack[$needle] - redaxo/src/core/lib/request.php:211:35
            return rex_type::cast($haystack[$needle], $vartype);

  call to rex_type::cast - redaxo/src/core/lib/request.php:211:35
            return rex_type::cast($haystack[$needle], $vartype);

  rex_type::cast#1 - redaxo/src/core/lib/util/type.php:44:33
    public static function cast($var, $vartype)

  $var - redaxo/src/core/lib/util/type.php:44:33
    public static function cast($var, $vartype)

  rex_type::cast - redaxo/src/core/lib/util/type.php:44:28
    public static function cast($var, $vartype)

  rex_request::arrayKeyCast - redaxo/src/core/lib/request.php:204:29
    private static function arrayKeyCast(array $haystack, $needle, $vartype, $default = '')

  rex_request::request - redaxo/src/core/lib/request.php:53:28
    public static function request($varname, $vartype = '', $default = '')

  rex_request - redaxo/src/core/functions/function_rex_globals.php:64:10
function rex_request($varname, $vartype = '', $default = '')

  rex_list::getSortColumn - redaxo/src/core/lib/list.php:827:21
    public function getSortColumn($default = null)

  $sortColumn - redaxo/src/core/lib/list.php:742:13
            $sortColumn = $this->getSortColumn();

  $params - redaxo/src/core/lib/list.php:744:17
                $params['sort'] = $sortColumn;

  $params - redaxo/src/core/lib/list.php:745:17
                $params['sorttype'] = $this->getSortType();

  array-fetch - redaxo/src/core/lib/list.php:750:18
        foreach ($params as $name => $value) {

  $value - redaxo/src/core/lib/list.php:750:38
        foreach ($params as $name => $value) {

  call to rex_list::replaceVariables - redaxo/src/core/lib/list.php:756:59
                $_params[$name] = $this->replaceVariables((string) $value);

  rex_list::replaceVariables#1 - redaxo/src/core/lib/list.php:926:38
    public function replaceVariables($value)

  $value - redaxo/src/core/lib/list.php:926:38
    public function replaceVariables($value)

  rex_list::replaceVariables - redaxo/src/core/lib/list.php:926:21
    public function replaceVariables($value)

  $columnValue - redaxo/src/core/lib/list.php:1136:21
                    $columnValue = $this->replaceVariables($columnValue);

  concat - redaxo/src/core/lib/list.php:1137:27
                    $s .= '        ' . $columnValue . "\n";

  concat - redaxo/src/core/lib/list.php:1137:27
                    $s .= '        ' . $columnValue . "\n";

  concatop - redaxo/src/core/lib/list.php:1137:21
                    $s .= '        ' . $columnValue . "\n";

  $s - redaxo/src/core/lib/list.php:1137:21
                    $s .= '        ' . $columnValue . "\n";

  concatop - redaxo/src/core/lib/list.php:1139:17
                $s .= '            </tr>' . "\n";

  $s - redaxo/src/core/lib/list.php:1139:17
                $s .= '            </tr>' . "\n";

  concatop - redaxo/src/core/lib/list.php:1143:13
            $s .= '        </tbody>' . "\n";

  $s - redaxo/src/core/lib/list.php:1143:13
            $s .= '        </tbody>' . "\n";

  concatop - redaxo/src/core/lib/list.php:1148:9
        $s .= '    </table>' . "\n";

  $s - redaxo/src/core/lib/list.php:1148:9
        $s .= '    </table>' . "\n";

  concatop - redaxo/src/core/lib/list.php:1149:9
        $s .= '</form>' . "\n";

  $s - redaxo/src/core/lib/list.php:1149:9
        $s .= '</form>' . "\n";

  rex_list::get - redaxo/src/core/lib/list.php:1022:21
    public function get()

  call to echo - redaxo/src/core/lib/list.php:1156:14
        echo $this->get();



------------------------------
1 errors found
``` 

Da verknüpft psalm zwei Verwendungsstellen, die gar nicht zusammenhängen. An der einen Stelle gibt es eine unsichere Eingabe, aber keine problematische Ausgabe.
An der anderen Stelle gibt es sichere Paramter, aber Ergebnis landet dann in Ausgabe.
Und psalm meldet nun die unsichere Eingabe der einen Stelle verknüpft mit der unsicheren Ausgabe der anderen Stelle.

Ich hatte verschiedene Dinge probiert, aber habe es nur so hinbekommen.

Reply:
hmm dann sollten wir das im kommentar für diesen phpdoc festhalten.
vermutlich lässt es sich nur schwer/nicht reprodzieren?
oder meinst du es macht sinn dass man diesen stand als separaten branch extrahiert und als bug in psalm meldet?

Reply:
Ich glaube, das ist kein Bug, sondern einfach das normale Verhalten der taint-analyse.
Es ist eigentlich recht leicht reproduzierbar: https://psalm.dev/r/e7cf937027

Nur mit specialize betrachtet psalm die Aufrufe alle separat.